src/bp/nlu/stan/model-repo.ts

```diff
@@ -65,16 +123,52 @@ export default class ModelRepository {
       ['model']
     )
     const buffer = await fse.readFile(archiveName)
-    const fpath = path.join(modelDir, modelFileName)
-    await fse.writeFile(fpath, buffer)
+    await scopedGhost.upsertFile(MODELS_DIR, modelName, buffer)
     tmpDir.removeCallback()
   }
 
-  private _makeFileName(modelId: NLUEngine.ModelId, password: string): string {
+  // // implement this properly
```

Comment:
We don't actually need a `listModels` method anyway.

What we truely need is a `hasModel(modelId, appId, appSecret)` which is secure.

And a `deleteModels(models: { modelId, appId, appSecret }[])` which is also secure.

I proposed a listModels method because it seems more generic and all pruposes, but if it's too complicated, let's go with a simpler solution.

Reply:
Also, here's another solution for the `listModels` method:

- instead of hashing `modelId` + `appSecret`, we use an encryption algo to make the file name
- when reading, we decrypt the file name with the given secret
- if the result of the decryption is valid `modelId` (we can check this with `modelIdService.isId(modelId)`) then we know the secret was ok.

Reply:
Yee but that requires to decrypt all models to validate the ownership (i know there won't be a ton). What I suggest is technically faster because you compute.the signature once and then check of that signature matches.

I didn't want to implement it because chances are YAGNI

Reply:
ok so what you suggest is to make the model file name something like:

```ts
const stringId = modelIdService.toString(modelId)
const appSecretHash = hash(`${appId}.${appSecret}`)
const modelFileName = `${appHash}.${stringId}` // appSecretHash.contentHash.specHash.seed.lang.model

// when listing all models of directory ${appId} with secret ${appSecret}

const allFiles = this.listDir(appId)

const expectedPrefix = hash(`${appId}.${appSecret}`)
const releventFiles = allFiles.filter(f => f.startsWith(expectedPrefix))
return releventFiles
```